File: backend/app/services/audit.py
# -*- coding: utf-8 -*-
"""系统业务操作审计，集中处理渠道、操作者与敏感参数脱敏。"""
from __future__ import annotations

from contextvars import ContextVar, Token
import json

from .. import db
from . import class_context

_actor: ContextVar[tuple[str, str]] = ContextVar(
    'workbench_actor', default=('web', 'local-user'))
_request_recorded: ContextVar[bool] = ContextVar('workbench_audit_recorded', default=False)
_SENSITIVE = (
    'key', 'token', 'secret', 'password', 'credential', 'authorization',
    '密码', '电话', '手机', '地址', '住址',
)


def bind_actor(channel: str | None, actor_id: str | None) -> Token:
    return _actor.set((str(channel or 'web')[:30], str(actor_id or 'local-user')[:80]))


def reset_actor(token: Token):
    _actor.reset(token)


def current_actor() -> tuple[str, str]:
    return _actor.get()


def begin_request() -> Token:
    return _request_recorded.set(False)


def reset_request(token: Token):
    _request_recorded.reset(token)


def has_recorded() -> bool:
    return _request_recorded.get()


def _sanitize(value, key: str = ''):
    if any(marker.lower() in key.lower() for marker in _SENSITIVE):
        return '***'
    if isinstance(value, dict):
        return {str(k): _sanitize(v, str(k)) for k, v in value.items()}
    if isinstance(value, list):
        return [_sanitize(item) for item in value[:20]]
    text = str(value) if value is not None else ''
    return text[:200]


def record(
    object_type: str, object_id: str | int = '', action: str = '', *,
    status: str = 'success', summary: str = '', params: dict | None = None,
    class_id: int | None = None, term_id: int | None = None, conn=None,
    commit: bool = True,
):
    _request_recorded.set(True)
    conn = conn or db.get_conn()
    if class_id is None or term_id is None:
        try:
            class_id, term_id = class_context.scope_ids(conn=conn)
        except class_context.ScopeError:
            class_id = term_id = None
    channel, actor_id = _actor.get()
    conn.execute(
        '''INSERT INTO system_audit(
               channel, actor_id, object_type, object_id, action, status,
               summary, params_summary, class_id, term_id
           ) VALUES(?,?,?,?,?,?,?,?,?,?)''',
        (channel, actor_id, object_type, str(object_id), action, status,
         str(summary or '')[:300], json.dumps(_sanitize(params or {}), ensure_ascii=False),
         class_id, term_id),
    )
    if commit:
        conn.commit()


def list_audits(limit: int = 200, conn=None) -> list[dict]:
    conn = conn or db.get_conn()
    class_id, term_id = class_context.scope_ids(conn=conn)
    rows = conn.execute(
        '''SELECT * FROM system_audit
           WHERE (class_id=? AND term_id=?) OR class_id IS NULL
           ORDER BY id DESC LIMIT ?''',
        (class_id, term_id, max(1, min(int(limit), 500))),
    ).fetchall()
    result = []
    for row in rows:
        item = dict(row)
        try:
            item['params_summary'] = json.loads(item['params_summary'])
        except (TypeError, ValueError):
            pass
        result.append(item)
    return result

File: server/src/services/audit.ts
/* MIG-04 系统审计：渠道/操作者上下文、敏感参数脱敏、缺省写审计。
 * 列表截断 20 项、字符串截断 200 字符。
 */
import { AsyncLocalStorage } from 'node:async_hooks';
import type { Database } from 'better-sqlite3';

import { getDb } from './context.js';

const actorStore = new AsyncLocalStorage<{ channel: string; actorId: string }>();
const recordedStore = new AsyncLocalStorage<boolean>();

const SENSITIVE_MARKERS = [
  'key', 'token', 'secret', 'password', 'credential', 'authorization',
  '密码', '电话', '手机', '地址', '住址',
];

export interface Actor {
  channel: string;
  actorId: string;
}

export function bindActor(channel: string | null, actorId: string | null): void {
  actorStore.enterWith({
    channel: String(channel ?? 'web').slice(0, 30),
    actorId: String(actorId ?? 'local-user').slice(0, 80),
  });
}

export function resetActor(): void {
  actorStore.enterWith({ channel: 'web', actorId: 'local-user' });
}

export function currentActor(): Actor {
  return actorStore.getStore() ?? { channel: 'web', actorId: 'local-user' };
}

export function beginRequest(): void {
  recordedStore.enterWith(false);
}

export function resetRequest(): void {
  recordedStore.enterWith(false);
}

export function hasRecorded(): boolean {
  return recordedStore.getStore() === true;
}

function sanitize(value: unknown, key = ''): unknown {
  if (SENSITIVE_MARKERS.some((marker) => key.toLowerCase().includes(marker.toLowerCase()))) {
    return '***';
  }
  if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
    const result: Record<string, unknown> = {};
    for (const [innerKey, innerValue] of Object.entries(value as Record<string, unknown>)) {
      result[String(innerKey)] = sanitize(innerValue, String(innerKey));
    }
    return result;
  }
  if (Array.isArray(value)) {
    return value.slice(0, 20).map((item) => sanitize(item));
  }
  /* Python str() 的布尔值为 True/False；JS String(false) 为 'false'，对齐 Python。 */
  const text = value === null || value === undefined ? ''
    : typeof value === 'boolean' ? (value ? 'True' : 'False') : String(value);
  return text.slice(0, 200);
}

export interface AuditOptions {
  status?: string;
  summary?: string;
  params?: Record<string, unknown> | null;
  classId?: number | null;
  termId?: number | null;
  conn?: Database;
  commit?: boolean;
}

export function record(
  objectType: string,
  objectId: string | number = '',
  action = '',
  options: AuditOptions = {},
): void {
  const conn = options.conn ?? getDb().connInstance;
  recordedStore.enterWith(true);
  let classId = options.classId ?? null;
  let termId = options.termId ?? null;
  if (classId === null || termId === null) {
    try {
      [classId, termId] = contextModule.scopeIds({ conn });
    } catch {
      classId = null;
      termId = null;
    }
  }
  const { channel, actorId } = currentActor();
  conn.prepare(
    'INSERT INTO system_audit('
    + 'channel, actor_id, object_type, object_id, action, status, '
    + 'summary, params_summary, class_id, term_id'
    + ') VALUES(?,?,?,?,?,?,?,?,?,?)',
  ).run(
    channel, actorId, objectType, String(objectId), action,
    options.status ?? 'success', String(options.summary ?? '').slice(0, 300),
    JSON.stringify(sanitize(options.params ?? {})), classId, termId,
  );
  // better-sqlite3 语句自动提交；commit 参数仅保留 API 兼容语义。
  void options.commit;
}

// 循环依赖处理：audit → context（scopeIds）；context 不依赖 audit。
import * as contextModule from './context.js';

export function listAudits(limit = 200, conn?: Database): Array<Record<string, unknown>> {
  const db = conn ?? getDb().connInstance;
  const [classId, termId] = contextModule.scopeIds({ conn: db });
  const rows = db.prepare(
    'SELECT * FROM system_audit '
    + 'WHERE (class_id=? AND term_id=?) OR class_id IS NULL '
    + 'ORDER BY id DESC LIMIT ?',
  ).all(classId, termId, Math.max(1, Math.min(Number(limit), 500))) as Array<Record<string, unknown>>;
  const result: Array<Record<string, unknown>> = [];
  for (const row of rows) {
    const item = { ...row };
    try {
      item.params_summary = JSON.parse(String(item.params_summary ?? '{}'));
    } catch {
      // 保持原样
    }
    result.push(item);
  }
  return result;
}
